File: java/src/sorter/StorageProviderTest.java
package sorter;

import java.io.File;
import java.io.IOException;

import org.junit.Assert;
import org.junit.jupiter.api.Test;

class StorageProviderTest {

    public void doTest(int recordSize, long recordCount, int stagingLength, int bufferLength) throws IOException {
        RandomGenerator rg = new RandomGenerator(recordCount, recordSize);
        Witness before = new Witness(rg);
        
        File storageFile = File.createTempFile("storage", ".tmp");
        storageFile.deleteOnExit();
        IoDevice storage = new IoDevice(storageFile);
        
        long storageOffset = 0;
        while(true) {
            Record rec = before.next();
            if(rec==null) break;
            storage.write(storageOffset, rec.data, 0, rec.data.length);
            storageOffset += rec.data.length;            
        }
                
        byte[] memory = new byte[10*1024*1024]; // 10MB
              
        
        long storageStartOffset = 0; // we wrote the records at offset zero        
   
        byte[] buffer = memory;
        int bufferStartOffset = 20;

        
        StorageProvider sp = new StorageProvider(
                recordSize,
                recordCount,
                storage,
                storageStartOffset,
                buffer,
                bufferStartOffset,
                bufferLength);
        
         
        Witness after = new Witness(sp);
        Consumer c = new Consumer(after);
        c.consume();
        
        Assert.assertEquals(before.getCount(), after.getCount());
        Assert.assertEquals(before.getCrc(), after.getCrc());
     }
    
    

    @Test
    void testSmall() throws IOException {
        doTest(123,50,564,2048);
     }

    @Test
    void testMedium() throws IOException {
        doTest(12003,50,1024,2048);
     }

}

File: Testing/Tests/StorageProviderTest.cpp
#include "StorageProviderTest.h"


StorageProviderTest::StorageProviderTest() {};

void StorageProviderTest::doTest(int recordSize, long recordCount, int stagingLength, int bufferLength) {

    RandomGenerator rg(recordCount, recordSize);
    Witness before(rg);

    IODevice storage("storage.tmp");

    long storageOffset = 0;
    while(true) {
        Record* recPtr = before.next();
        if(!recPtr) break;
        Record rec = *recPtr;
        storage.write(storageOffset, rec.data, 0, sizeof(rec.data));
        storageOffset += sizeof(rec.data);
    }

    char * memory = new char [10*1024*1024]; // 10MB


    long storageStartOffset = 0; // we wrote the records at offset zero

    char * buffer = memory;
    int bufferStartOffset = 20;
    uint32_t keyOffset = 8;


    StorageProvider sp(
            recordSize,
            recordCount,
            storage,
            storageStartOffset,
            buffer,
            bufferStartOffset,
            bufferLength, keyOffset);


    Witness after(sp);
    Consumer c(after);
    c.consume();

    assert(("The count of the before witness should have equaled the count of the after witness" && before.getCount() == after.getCount()));
    assert(("The checksum of the before witness should have equaled the checksum of the after witness" && before.getCrc() == after.getCrc()));

    remove("storage.tmp");
}

void StorageProviderTest::testSmall() {
    doTest(123,50,564,2048);
}

void StorageProviderTest::testMedium() {
    doTest(123,50,564,2048);
}

int main(){
    StorageProviderTest test;
    test.testSmall();
    test.testMedium();
}